README.md
# Introduction
RandomVariates is a library of random variate generation routines.
The purpose behind this library was purely for educational purposes as 
a way to learn how to generate random variates using such methods as 
inverse transform, convolution, acceptance-rejection and composition 
methods. Additionally, this project was an excuse to get familiar with 
random number generators such as linear congruential generators, 
Tausworthe Generators and Widynski's "Squares: A Fast Counter-Based RNG"

## Pseudo Random Number Generators
The following pseudo random number (PRN) generators are contained in this project:
* A basic "desert island" linear congruential (implemented in the uniform function)
* taus() and tausunif(): A basic Tausworthe PRN generator and a Tausworthe Uniform 
PRN generator
* squaresrng(): Widynski's "Squares: A Fast Counter-Based RNG" 
https://arxiv.org/pdf/2004.06278.pdf 

### Various helper functions to take advantage of the PRN generators
* randseed(): Helper function to grab a "smaller" PRN from the Widynski squares PRN 
generator
* generateseed(): Helper function to generate random seeds if the initial seed has 
not been set
* set_seed() and get_seed(): Functions to get and set the seed.
* reverse(): Helper function to reverse an integer 

## Random Variate Generation Routines
* uniform(): Routine to generate uniform random variates between a and b. 
Default uniform(a=0, b=1)
* norm(): Method to generate random normals. Default norm(mu=0, sd=1)
* exponential(): Generate exponential random variates. 
Default exponential(lam=1)
* erlang(): Routine to generate Erlang_k(lambda) random variates. 
Default erlang(lam=1, k=1, n=1)
* weibull(): Method to generate weibull random variates: Default 
weibull(lam=1, beta=1)
* triangular() Generate triangular random variates with a-lower, b-mode 
and c-upper Default triangular(a=0, b=1, c=2)
* Bernoulli(p) random variates Default bernoulli(p=0.5)
* Routine to generate binomial random variates Default binomial(t=1, p=0.5)
* dicetoss() Simple/fun method to generate X-sides dice toss. Default is 
a simple 6-sided dicetoss(sides=6)
* geometric() Method to generate geometric random variates 
Default geometric(p=0.5)
* negbin() Routine to generate discrete random negative binomials 
Default negbin(t=1, p=0.5)
* chisq() Generate Chi-squared random variates Default chisq(df=1)
* poisson() Method to generate Poisson random variates Default poisson(lam=1)
* gamma() Gamma random variates shape parameter k and a scale parameter θ. 
Implementation is based on Marsaglia and Tsang's transformation-rejection method
of generating gamma random variates 
(https://dl.acm.org/doi/10.1145/358407.358414) Default gamma(k=1.0, theta=1)
* lognormal() Generate lognormal random variates Default lognormal(mu=0, sd=1)
* beta() Routine to generate beta random variates Default beta(a=1, b=1)

## Limitations
* Unlike Numpy's random variate generation routines, these are written
in python. Numpy's routines are written in C hence are much, much faster.
* Beta and Gamma distributions only accept a, b, k and theta greater than one. 
Other random variate implementations, such as Numpy can handle values between
0 and 1.
* Setting the seed does not affect the Tausworthe and Tausworthe Uniform PRN 
generators

### Distributions not currently implemented
* Pearson Type V
* Pearson Type VI
* Log-Logistic
* Johnson Bounded and Johnson unbounded
* Bézier

## Installation
### Requirements:
* Python 3.x
* pip (https://pip.pypa.io/en/stable/installation/)

To install the library, simply run the command:
* pip install randvars

## Usage
To use the library, you need to import the library into your python script then 
create an instance of random variates:
> import randomvariates

> rv = randomvariates.random.RandomVariates()

Alternately you can import random from randomvariates:
> from randomvariates import random

> rv = random.RandomVariates()

### Seeds
By default, a seed is not set when an instance or randomvariates is called.
When a seed is set to None, randomvariates will randomly generate values for the
various random variate routines. For repeatability, we can set a seed by calling 
the set_seed() method. Once a seed has been set, we can verify by calling the 
get_seed() method.

> from randomvariates import random
>
> rv = random.RandomVariates()
> 
> rv.set_seed(42)
>
> rv.get_seed()
> 
> 42

### Pseudo Random Number Generators
To call the Widynski Squares PRN we can call the squaresrng() method. 
The squaresrng() method takes a center and key value. By default, the center
and key are set to 1: squaresrng(ctr=1, key=1)
> rv.squaresrng(42,21)
> 
> 22904061750312427071608663841693658494663185320788517623007713567980053732104718807902410691731255108163475339984462249791973853173096390867949739437289512015166556428304384

As of 11-06-2021, the Tausworthe PRN  and Tausworthe Uniform PRN generator does 
not take a seed value (See Limitations above)
To call the Tausworthe generators, simply call rv.taus() and rv.tausunif(). 
By default taus() will generate 100 binary PRNs and rv.tausunif() will generate 
a single uniform(0,1):

>rv.taus(n=100)
> 
> array([0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1,
       0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0])

> rv.tausunif(n=1)
> 
>array([0.22627192])

**Linear Congruential Generator (LCG)**

The Uniform PRN generator is based off a "desert island" LCG of the form:

> X_{i} = 16807 X_{i-1} mod(2**32 - 1)

To call the uniform PRN generator simply call:
> rv.uniform()
> 
> array([0.0028378])

To generate more than one unif(0,1), call the method with n=X where X is the number 
of unif(0,1)s to generate:

> rv.uniform(n=25)
> 
> array([0.0028378 , 0.69495865, 0.17008364, 0.59578035, 0.28035944,
       0.00113405, 0.05993272, 0.28927888, 0.91019703, 0.68147951,
       0.62609725, 0.81650574, 0.01200872, 0.83049583, 0.14336138,
       0.4747437 , 0.01741077, 0.62288295, 0.79372406, 0.12022883,
       0.68598241, 0.3064384 , 0.31021374, 0.76239295, 0.53823463])

If we want to generate something other than unif(0,1), we can call the function
with a=X and b=Y where X and Y are the lower and upper bounds of the uniform 
distribution:

> rv.uniform(a=7, b=11, n=25)
> 
> array([ 7.01135121,  9.77983461,  7.68033457,  9.3831214 ,  8.12143777,
        7.00453619,  7.23973089,  8.15711553, 10.64078812,  9.72591803,
        9.50438901, 10.26602297,  7.04803487, 10.32198331,  7.57344553,
        8.89897481,  7.0696431 ,  9.4915318 , 10.17489623,  7.48091533,
        9.74392966,  8.22575361,  8.24085498, 10.04957178,  9.15293853])

### Distributions
**Normal Random Variates**

To generate random normal random variates, call the norm() function. By default, 
the norm() function will generate values with mean = 0 and standard deviation = 1.

> rv.norm(n=25)
> 
> array([-1.33438863,  0.12180611,  0.88656523,  0.50965537, -1.64358406,
       -0.25778164,  0.57095618,  1.90310886, -0.05967737, -0.34183211,
        1.40942348,  0.588753  , -2.00879407, -0.27557057, -0.05367554,
        0.36562436,  1.51957859, -0.87597507,  0.27341912,  0.99870143,
        0.0563413 , -0.58931763,  0.06256761,  1.34552544, -0.41456673])

To generate normals with other means and standard deviations, simply specify them 
when calling the function:

> rv.norm(mu=42, sd=21, n=25)
> 
> array([31.09197496, 90.91916642, 10.96438887, 63.22805106, 11.65331438,
       42.3934924 , 31.50241102, 57.32494887, 39.63622134, 50.84789244,
       29.66813461, 71.59768198, 51.23679519, 29.62926174, 38.93133399,
       21.33704934, 44.01056639, 85.43369206, 10.93161744, 35.5352881 ,
       47.6567116 , 62.89812129, 35.67247842, 48.76775665, 37.78179072])

**Exponential Random Variates**

By default, the exponential() functin will generate a single, lambda=1 random variate.

> rv.exponential()
> 
> array([1.11685025])

To generate exponentials with different rates (lambda), call the exponential function
with lam=X, where X is 1/X rate:

> rv.exponential(lam=42, n=25)
> 
> array([0.02659167, 0.05889761, 0.0237974 , 0.05500891, 0.00990581,
       0.00665694, 0.01027128, 0.00311863, 0.00991187, 0.00155661,
       0.02532011, 0.00358105, 0.06095103, 0.02608557, 0.03126096,
       0.01844892, 0.04179501, 0.00204212, 0.01357267, 0.02956585,
       0.05493093, 0.04847486, 0.03083947, 0.03648715, 0.02353137])

**Erlang Random Variates**

Random Erlang variates can be generated by calling the erlang() function. By default,
the erlang function will generate variates with lambda = 1 and shape (k) = 1:

> rv.erlang()
> 
> array([0.39646936])

To generate erlangs with different rate and shape parameters, set lam=X and k = Y,
where X is the lambda rate and Y is the shape:

> rv.erlang(lam=5, k=5, n=25)
> 
> array([0.39646936, 0.08803646, 0.45897164, 0.10449879, 1.07778402,
       1.41096694, 1.04868842, 2.09746919, 1.07729126, 2.76008462,
       0.42352393, 1.96869049, 0.08046101, 0.40697774, 0.31337257,
       0.61761435, 0.18975684, 2.49867301, 0.8335498 , 0.34090517,
       0.10486022, 0.13990266, 0.31996746, 0.24335083, 0.46553741])

**Weibull Random Variates**

To generate values from the Weibull distribution, call the weibull() method with 
lam and beta. By default, lam and beta are set to 1 weibull(self, lam=1, beta=1).

> rv.weibull()
> 
> array([0.00284184])

To generate weibull values with different lam (shape) and beta (scale), set lam 
and beta as such:

> rv.weibull(lam=3, beta=5, n=25)
> 
> array([0.10318073, 0.34497802, 0.23822211, 0.32680215, 0.26688172,
       0.08587253, 0.19102132, 0.26887509, 0.39745421, 0.3424279 ,
       0.33224353, 0.37046057, 0.13781682, 0.37386295, 0.2295085 ,
       0.30523775, 0.14852653, 0.33166333, 0.36519926, 0.22099585,
       0.34327599, 0.2726182 , 0.2734267 , 0.35840856, 0.31657801])

**Triangular Random Variates**

By default, the randomvariates library will generate Triangular(0,1,2) values from 
a triangular distribution:

> rv.triangular()
> 
>array([0.07533662])

To generate values from a Triangular distribution with lower bound a, mode b and 
upper bound c, call the triangular() function with a, b, and c set:

> rv.triangular(a=-5, b=0, c=5, n=25)
> 
> array([-4.6233169 ,  1.09461047, -2.08380691,  0.50433737, -1.25594176,
       -4.76187742, -3.26892052, -1.19685075,  2.88100295,  1.00925764,
        0.67621261,  1.97102115, -4.22512206,  2.08877884, -2.32267501,
       -0.12791778, -4.06697336,  0.65766738,  1.7884899 , -2.54817586,
        1.0375665 , -1.08567757, -1.06163902,  1.55321124,  0.19497467])

**Bernoulli Random Variates**

To generate bernoulli(p) random values, call the bernoulli() method with 
probability, p. By default, the bernoulli() method generates bernoulli(0.5) random
values.

> rv.bernoulli()
> 
> array([0])
> 
> rv.bernoulli(n=25)
> 
>array([0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0,
       0, 1, 1])

To generate bernoulli(0.8) random values, set p=0.8:

> rv.bernoulli(p=0.8, n=25)
> 
> array([0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1,
       1, 1, 1])

**Binomial Random Variates**

Binomial(n,p) random values can be generated with the binomial() function.
By default, the binomial() function generates 1 trial at p=0.5:

> rv.binomial()
> 
> array([0])

Note: Don't confuse t=trials and n=number of values to generate. To generate 25 
binomials with 10 trials, and probability 0.5, we would specify 
binomial(t=10, p=0.5, n=25): 

> rv.binomial(t=10, p=0.5, n=25)
> 
> array([6, 5, 3, 3, 5, 3, 5, 7, 6, 5, 3, 5, 4, 5, 4, 4, 2, 6, 6, 6, 4, 4,
       6, 5, 2])

**Random X-sided Dice Toss**

For the D&D fans, the dicetoss() function allows you to generate an X-sided
die toss. For example, to generate 10, 20-sided dice tosses, simply call the 
dicetoss() function. By default, dicetoss() defaults to a 6-sided die:

> rv.dicetoss(n=10)
> 
> array([6., 6., 5., 1., 6., 3., 1., 5., 3., 5.])

To generate 10, 20-sided dice toss, set the side variable to 20:

>rv.dicetoss(sides=20, n=10)
> 
> 
> array([20., 20., 17.,  4., 20., 10.,  3., 15., 10., 14.])

**Geometric Random Variates**

To generate geometric random values, use the geometric() function. By default, 
the geometric function is set to a probability of 0.5:

> rv.geometric()
> 
> array([5.])

To generate geometric values with a different probability, set p equal to the new 
probability:

> rv.geometric(p=0.42, n=25)
> 
> array([ 7., 12.,  4.,  1.,  6.,  2.,  1.,  3.,  2.,  3.,  1.,  1.,  2.,
        1.,  2.,  7.,  3.,  1.,  4.,  2.,  1.,  2.,  2.,  4.,  1.])



HISTORY.md
# Release History

2021-11-10: More stupid seed fixes

2021-11-10: Erlang fixes

2021-11-08: Beta fixes

2021-11-08: Gamma fixes

2021-11-08: Refactored poisson to correctly handle the situation when a seed is set

2021-11-08: Found a bug in the seed algorithm

2021-11-08: Refactored norm to properly hand seed==0

2021-11-08: Refactored chisq method to work around "funky" seeding issues

2021-11-08: Refactored the negbin method to properly handle the situation when a
seed is set.

2021-11-08: Refactored the binomial method to properly handle the situation when a
seed is set.

2021-11-06: Initial creation of pip package

2021-11-06: Added Tausworthe Generator and a uniform generator based off the 
Tausworthe generator.

randomvariates/random.py
"""
Author: John Paul Goodman
Course: ISyE 6644
Semester: Fall 2021
Project Group: 40
"""

import numpy as np
from datetime import datetime


class RandomVariates:
    """
    Class method to generate random variates
    """

    def __init__(self):
        self.seed = None
        self.prn = 0
        self.seed0 = 2 ** 23 - 1
        self.m = 16807
        self.m31 = (2 ** 31 - 1)
        self.__version__ = "0.0.17"

    @staticmethod
    def reverse(n=0):
        """
        reverses an integer
        :return: reversed integer
        """
        rev = 0
        while n > 0:
            rev = rev * 10 + n % 10
            n = n // 10
        return rev

    @staticmethod
    def squaresrng(ctr=1, key=1):
        """
        B. Widynski
        https://arxiv.org/pdf/2004.06278.pdf
        :param ctr: the center
        :param key: the key
        :return: very large random number
        """
        y = x = ctr * key
        z = y + key
        x = x * x + y
        x = (x >> 32) | (x << 32)  # round 1
        x = x * x + z
        x = (x >> 32) | (x << 32)  # round 2
        x = x * x + y
        x = (x >> 32) | (x << 32)  # round 3
        return (x * x + z) >> 32  # round 4

    def randseed(self):
        """
        Get random number from square
        :return: smaller random number
        """
        i = int(datetime.now().timestamp() * 1000000)
        r = self.reverse(int(datetime.now().timestamp() * 1000000))
        self.prn = self.squaresrng(i, r) % 100000000000
        return self.prn

    def taus(self, n=100):
        """
        Basic Tausworthe Random Number Generator
        :param n: number of random binary values to return
        :return:
        """
        rseed = self.randseed()
        bstring = format(rseed, 'b')
        b = [int(x) for x in bstring]
        blen = len(b)
        r = int(np.round(blen * 0.25))
        q = int(np.round(blen * 0.75))
        for _ in range(n):
            bi = b[blen - r] ^ b[blen - q]
            b.append(bi)
            blen = len(b)
        diff = len(b) - n
        return np.array(b[diff:])

    def tausunif(self, n=1):
        """
        Generate uniform PRNs based off the Tausworthe PRN
        :param n: number of uniform PRNs to generate
        :return: a numpy array of uniform PRNs
        """
        m = 100
        x = self.taus(n=n * m)
        x = [str(i) for i in x]
        it = [iter(x)] * m
        obs = np.array([int(''.join(s), 2) / (2 ** m) for s in zip(*it)])
        return np.array(obs)

    def set_seed(self, seed):
        """
        Set the seed
        :param seed: seed value
        """
        self.seed = seed

    def get_seed(self):
        """
        Get the seed
        """
        return self.seed

    def generateseed(self):
        """
        Generate a random seed if seed value is None. This helps with randomly generating values.
        :return: a seed value
        """
        if self.seed is None:
            self.seed0 = self.randseed()
            seed = (self.m + self.seed0) % self.m31
        else:
            # This is a bug self.seed0 doesn't get unset. Get rid of it from this calculation
            # seed = ((self.m + self.seed0 + self.seed) % self.m31) ** np.pi
            seed = ((self.m + self.seed) % self.m31) ** np.pi
        return seed

    def uniform(self, a=0, b=1, n=1):
        """
        Generate uniform random variates
        :param n: number of uniform RVs to generate
        :param a: starting point of uniform range
        :param b: ending point of uniform range
        :return: a numpy array of uniform RVs
        """
        unifs = [self.generateseed()]  # type: list[int]
        for i in range(1, n + 1):
            xi = (self.m * unifs[i - 1]) % self.m31
            unifs.append(xi)
        # simple list append is faster than numpy.append()
        # numpy.append() is shockingly slow. this is because it does a full copy?
        return np.array([(((x * (b - a)) / self.m31) + a) for x in unifs][1:])

    def norm(self, mu=0, sd=1, n=1):
        """
        Generate Normal Random Variates
        :param mu: mean
        :param sd: standard deviation
        :param n: number of random normals to generate
        :return: a numpy array of random normals
        """
        seed = self.seed  # grab init seed
        u1 = self.uniform(n=n)
        if self.seed or self.seed == 0:  # is not None:
            self.seed += 2 ** 34 - 1  # Hack to make sure U1 and U2 look independent
        u2 = self.uniform(n=n)
        theta = 2 * np.pi * u2
        r = np.sqrt(-2 * np.log(u1))
        x = r * np.cos(theta)
        z = mu + (x * sd)
        self.seed = seed  # reset init seed
        return z

    def exponential(self, lam=1, n=1):
        """
        Generate Exponential Random Variates
        :param lam: lamba or rate
        :param n: the number of random variates to generate
        :return: a numpy array of random exponentials
        """
        if lam == 0 or lam == 0.0:
            return np.zeros(n)  # zeros for summation
        u1 = self.uniform(n=n)
        exp = (-1 / lam) * np.log(1 - u1)
        return exp

    def erlang(self, lam=1, k=1, n=1):
        """
        Generate Erlang Random Variates
        :param lam: lambda or rate
        :param k: shape parameter
        :param n: number of random variates
        :return: a numpy array of random erlang
        """
        seed = self.seed
        erl = np.ones(n)  # ones for product
        for i in range(k):
            erl *= self.uniform(n=n)
            if self.seed or self.seed == 0:
                self.seed += 1
        erl = (-1 / lam) * np.log(erl)
        self.seed = seed
        return erl

    def weibull(self, lam=1, beta=1, n=1):
        """
        Generate Weibull Random Variates
        :param lam: lambda aka the scale
        :param beta: beta aka the shape
        :param n: number of random variates to generate
        :return: a numpy array of random weibull variates
        """
        return (1 / lam) * (-1 * np.log((1 - self.uniform(n=n)))) ** (1 / beta)

    def triangular(self, a=0, b=1, c=2, n=1):
        """
        Generate Triangular Random Variates
        :param a: lower limit of the triangle
        :param b: mode or peak of the triangle
        :param c: upper limit of the triangle
        :param n: number of triangle random variates
        :return: a numpy array of random triangle variates
        """
        f_c = (b - a) / (c - a)
        u = self.uniform(n=n)
        x_tris = []
        for ui in u:
            if ui <= f_c:
                xi = a + np.sqrt(ui * ((b - a) * (c - a)))
                x_tris.append(xi)
            else:
                xi = c - np.sqrt((1.0 - ui) * ((c - b) * (c - a)))
                x_tris.append(xi)
        return np.array(x_tris)

    def bernoulli(self, p=0.5, n=1):
        """
        Generate Bernoulli(p) random variates
        :param p: probability of success
        :param n: number of "coin" tosses
        :return: a numpy array of random "coin" tosses
        """
        u = self.uniform(n=n)
        bp = []
        for ui in u:
            if ui <= 1 - p:
                bp.append(0)
            else:
                bp.append(1)
        return np.array(bp)

    def binomial(self, t=1, p=0.5, n=1):
        """
        Draw samples from a binomial distribution
        :param t: number of bernoulli(p) trials
        :param p: probability of success
        :param n: number of binomial variates to generate
        :return: a numpy array of samples from the parameterized binomial distribution, where each sample is equal to
        the number of successes over the n trials.
        :note: this is clearly not as fast as numpy
        """
        binomials = []
        count = 0
        for i in range(n):
            if self.seed:
                self.seed += i
                count += i
            bernsum = np.sum(self.bernoulli(p=p, n=t))
            binomials.append(bernsum)
        if self.seed or self.seed == 0:
            self.seed -= count
        return np.array(binomials)

    def dicetoss(self, sides=6, n=1):
        """
        Simple dice toss of die with x-side count of "sides" (i.e. 6, 10, 20, etc)
        :param sides: number of sides on the die
        :param n: number of dice tosses
        :return: a numpy array of x-sided dice tosses
        """
        u = self.uniform(n=n)
        return np.ceil(sides*u)

    def geometric(self, p=0.5, n=1):
        """
        Generate geometric random variates
        :param p: probability of success
        :param n: number of geometric random variates to generate
        :return: a numpy array of geometric random variates
        """
        u = self.uniform(n=n)
        return np.ceil(np.log(1-u)/np.log(1-p))

    def negbin(self, t=1, p=0.5, n=1):
        """
        Generate negative binomial random variates
        :param t: number of trials
        :param n: number of random variates to generate
        :param p: probability of success
        :return: a numpy array of random negative binomials
        :note: this is painfully slow for large n
        """
        negbins = []
        seed = self.seed
        for i in range(n):
            if self.seed or self.seed == 0:
                self.seed += i
            nbt = np.sum(self.geometric(p=p, n=t))
            negbins.append(nbt)
        self.seed = seed
        return np.array(negbins)

    def chisq(self, df=1, n=1):
        """
        Generate Chi square random variates with df degrees of freedom
        :param df: degrees of freedom
        :param n: number of random variates to produce
        :return: a numpy array of chi square random variates
        """
        seed = self.seed
        chi = np.zeros(n)
        for i in range(df):
            chi += self.norm(n=n) ** 2
            if self.seed or self.seed == 0:
                self.seed = seed + i
        self.seed = seed
        return chi

    def poisson(self, lam=1, n=1):
        """
        Generate Poisson random variates
        :param lam: lambda or rate
        :param n: number of random poissons to generate
        :return: a numpy array of random poissons
        """
        p = []
        count = 0
        seed = self.seed
        for _ in range(n):
            t = 0  # initialize sum of exponential variables as zero
            n = -1  # initialize counting variable as negative one
            while t < 1:
                e = self.exponential(lam=lam)[0]
                t += e
                n += 1
                count += 1
                if self.seed or self.seed == 0:
                    self.seed += 1
            p.append(n)
        self.seed = seed
        return np.array(p)

    def gamma(self, k=1.0, theta=1, n=1):
        """
        Instead of implementing the acceptance-rejection algorithms for generating gamma variates described by
        Law (2015), this is an implementation of Marsaglia and Tsang's transformation-rejection method of using one
        normal variate and one uniform variate as demonstrated in their 2000 paper:
        https://dl.acm.org/doi/10.1145/358407.358414
        :param k: shape parameter
        :param theta: scale parameter
        :param n: number of random gammas to generate
        :return: a numpy array of gamma random variates
        """
        assert k >= 1.0, "k should be at least 1.0"
        assert theta >= 1.0, "theta should be at least 1.0"
        seed = self.seed
        d = k - 1 / 3
        c = 1 / np.sqrt(9 * d)
        gammas = []
        count = 0
        while len(gammas) < n:
            if self.seed or self.seed == 0:
                self.seed += count
            x = self.norm()[0]
            u = self.uniform()[0]
            v = (1 + c * x) ** 3
            if (x > -1 / c) and (np.log(u) < (0.5 * (x ** 2) + d - d * v + d * np.log(v))):
                gammas.append(d * v)
            count += 1
        self.seed = seed
        return theta * np.array(gammas)

    def lognormal(self, mu=0, sd=1, n=1):
        """
        Lognormal Random Variates. The lognormal distribution occasionally referred to as the
        Galton distribution or Galton's distribution, after Francis Galton.
        :param mu: mean
        :param sd: standard deviation
        :param n: number of lognormals to generate
        :return: a numpy array of lognormal random variates
        """
        # 1) Generate Y ~ Nor(u, var)
        # 2) Return X = np.e**Y
        y = self.norm(mu=mu, sd=sd, n=n)
        return np.e**y

    def beta(self, a=1, b=1, n=1):
        """
        Generate beta random variates
        :param a: shape parameter a
        :param b: shape parameter b
        :param n: number of beta random variates to generate
        :return: a numpy array of beta random variates
        """
        # 1) Generate Y1 ~ gamma(a, 1) and Y2 ~ gamma(a, 1)
        # 2) Return X = Y1/(Y1+Y2)
        assert a >= 1.0, "a should be at least 1.0"
        assert b >= 1.0, "b should be at least 1.0"
        seed = self.seed
        y1 = self.gamma(k=a, theta=1, n=n)
        if self.seed:
            self.seed += 2 ** 17 - 1
        y2 = self.gamma(k=b, theta=1, n=n)
        x = y1 / (y1 + y2)
        self.seed = seed
        return np.array(x)
